### PR TITLE
Select the strongest nodes #329

### PR DESCRIPTION
**Pull request by botanicus**
This PR implements selecting the strongest nodes as described in #329.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#355](https://github.com/zold-io/zold/issues/355) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#355](https://github.com/zold-io/zold/issues/355) is assigned to @ledestin[/z](https://www.0crat.com/u/ledestin), here is [why](https://www.0crat.com/footprint/CAZPZR9FS/1871418039); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: test/commands/test_remote.rb [hunk 1]
```diff
@@ -116,4 +116,42 @@ def test_remote_trim_with_tolerate
       assert_equal(1, remotes.all.count)
     end
   end
+
+  # @todo #329:30min Verify that the nodes that are being selected are
+  #  really the strongest ones. The strongest nodes are the ones with
+  #  the highest score.
+  def test_select_selects_the_strongest_nodes
```
Comment (ledestin) [inline test/commands/test_remote.rb]: @botanicus maybe `test_strongest_nodes_are_selected`

File: test/commands/test_remote.rb [hunk 2]
```diff
@@ -116,4 +116,42 @@ def test_remote_trim_with_tolerate
...
+    skip
+  end
+
+  def test_select_respects_max_nodes_option
+    Dir.mktmpdir 'test' do |dir|
+      remotes = Zold::Remotes.new(File.join(dir, 'remotes.txt'))
+      zero = Zold::Score::ZERO
+      cmd = Zold::Remote.new(remotes: remotes, log: test_log)
+      (1..10).each do |int|
```
Comment (ledestin) [inline test/commands/test_remote.rb]: @botanicus The role of ` int` is `port_offset`, is it not?

File: test/commands/test_remote.rb [hunk 3]
```diff
@@ -116,4 +116,42 @@ def test_remote_trim_with_tolerate
...
+            ]
+          }.to_json
+        )
+        stub_request(:get, "http://localhost:#{5000 + int}/version").to_return(
+          status: 200,
+          body: {
+            version: Zold::VERSION
+          }.to_json
+        )
+        cmd.run(%W[remote add localhost #{5000 + int}])
```
Comment (ledestin) [inline test/commands/test_remote.rb]: @botanicus I'd say calculating `5000 + int` many times isn't DRY.

File: test/commands/test_remote.rb [hunk 4]
```diff
@@ -116,4 +116,42 @@ def test_remote_trim_with_tolerate
...
+      end
+      assert_equal(11, remotes.all.count)
+
```
Comment (ledestin) [inline test/commands/test_remote.rb]: Yegor will scold you for the empty line.

**Comment by ledestin** [comment]
@botanicus hey, check out my comments

File: test/commands/test_remote.rb (hunk 1 above)
Comment (botanicus) [reply test/commands/test_remote.rb]: What I’m trying to do here is to (sort of) namespace the subcommands. The `remote` command has many subcommands and currently we’re testing them in one Minitest class.

With this convention (`test_subcommandname_testdesc`) one can see which tests test which subcommand. I’m not saying it’s the best approach though, rather an ad-how one.

I think we need a better solution than that though, so I reported it here #360.

Comment (ledestin) [reply test/commands/test_remote.rb]: @botanicus got you. In RSpec you could say `describe '#select'`.

Comment (botanicus) [reply test/commands/test_remote.rb]: @ledestin exactly. I missed the `#`. Updated now, cheers!

File: lib/zold/commands/remote.rb [hunk 5]
```diff
@@ -251,10 +251,13 @@ def update(opts, deep = true)
       end
     end
 
-    # @todo #292:30min Implement the logic of selecting the nodes as per #292.
-    #  The strongest n nodes should be selected, where n = opts['max-nodes'].
-    def select(_opts)
-      raise NotImplementedError, 'This feature is not yet implemented.'
+    def select(opts)
+      max_nodes = opts['max-nodes']
```
[the remark this answers is not in the document]
Comment (botanicus) [reply lib/zold/commands/remote.rb]: @ledestin makes sense, but it’s not a hash, but a slop option object that doesn’t have `#fetch` – I checked.

**Comment by botanicus** [comment]
@ledestin should be all now.

**Comment by botanicus** [comment]
@ledestin can you please confirm that everything has been addressed?

**Comment by ledestin** [comment]
@botanicus sorry, I somehow missed it yesterday

File: test/commands/test_remote.rb [hunk 6]
```diff
@@ -116,4 +116,42 @@ def test_remote_trim_with_tolerate
...
+  def test_select_selects_the_strongest_nodes
+    skip
+  end
+
+  def test_select_respects_max_nodes_option
+    Dir.mktmpdir 'test' do |dir|
+      remotes = Zold::Remotes.new(File.join(dir, 'remotes.txt'))
+      zero = Zold::Score::ZERO
+      cmd = Zold::Remote.new(remotes: remotes, log: test_log)
+      (5000..5010).each do |port|
```
Comment (ledestin) [inline test/commands/test_remote.rb]: @botanicus how do you feel about extracting this to a method?

**Comment by ledestin** [comment]
@botanicus please see

File: test/commands/test_remote.rb (hunk 6 above)
Comment (botanicus) [reply test/commands/test_remote.rb]: @ledestin I'm feeling like I seriously run out of time on this one. It's been going on for 9 days and I only get half an hour paid.

**Comment by botanicus** [comment]
@ledestin please see. Can we close?

**Comment by 0crat** [comment]
@ledestin[/z](https://www.0crat.com/u/ledestin) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @ledestin[/z](https://www.0crat.com/u/ledestin), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#355](https://github.com/zold-io/zold/issues/355) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256)

**Comment by ypshenychka** [comment]
@botanicus According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>Messages in a job [always](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html#4-avoid-noiseaddress-your-comments) start with a name of a user they are addressed to.

Please correct your [message](https://github.com/zold-io/zold/pull/355#discussion_r200784840) by indicating an addressee in the beginning.

**Comment by ypshenychka** [comment]
@ledestin According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>Messages in a job [always](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html#4-avoid-noiseaddress-your-comments) start with a name of a user they are addressed to.

Please correct your [message](https://github.com/zold-io/zold/pull/355#discussion_r200782097) by indicating an addressee in the beginning.

**Comment by ledestin** [comment]
@ypshenychka done

**Comment by ypshenychka** [comment]
@ledestin thanks

**Comment by ypshenychka** [comment]
@botanicus ping

**Comment by botanicus** [comment]
@ypshenychka done

**Comment by ypshenychka** [comment]
@botanicus thanks

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @ledestin[/z](https://www.0crat.com/u/ledestin)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**